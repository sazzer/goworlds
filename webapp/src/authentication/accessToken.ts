import {createReducer} from "redux-create-reducer";
import {Action, BaseAction, buildAction} from "../redux";
import produce from "immer";
import {MODULE_PREFIX} from "./module";

/** The shape of an access token */
export declare type Token = {
    token: string,
    expires: Date,
};

/** The shape of the state for this module */
export declare type State = {
    token: Token | undefined,
};

/** The initial state for this module */
const INITIAL_STATE: State = {
    token: undefined,
};

//////// Action for checking if an email address exists

/** The action type for storing an access token */
const STORE_ACCESS_TOKEN_ACTION = MODULE_PREFIX + 'storeAccessToken';

/** The shape of the action to store an access token */
declare type StoreAccessTokenAction = {
    token: string,
    expires: Date,
}

/**
 * Build the Redux action to store an access token
 * @param token The actual token
 * @param expires When the token expires
 */
export function storeAccessToken(token: string, expires: Date): Action<StoreAccessTokenAction> {
    return buildAction(STORE_ACCESS_TOKEN_ACTION, {
        token,
        expires
    });
}

/**
 * Reducer for storing the access token
 * @param state the state to update
 * @param action the action to update the state with
 * @return the new state
 */
export function storeAccessTokenReducer(state: State, action: Action<StoreAccessTokenAction>) : State {
    return produce(state, (draft: State) => {
        draft.token = action.payload;
    });
}

//////// The actual module definitions

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
    [STORE_ACCESS_TOKEN_ACTION] : storeAccessTokenReducer as ((state: State, action: BaseAction) => State)
});
