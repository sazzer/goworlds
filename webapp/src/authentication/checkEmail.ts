import {from, Observable} from "rxjs";
import {request} from "../api";

/** The shape of the response from the server on checking if an email address exists */
declare type CheckEmailExistsServiceResponse = {
    exists: boolean,
};

/**
 * Check if the given Email Address is already registered
 * @param email the email address to check
 * @return the response of the check
 */
export function checkEmail(email: string) : Observable<boolean> {
    const response = request<CheckEmailExistsServiceResponse>('/emails/{email}', {
        method: 'GET',
        urlParams: {
            email,
        }
    })
        .then(response => response.body.exists);

    return from(response);
}
