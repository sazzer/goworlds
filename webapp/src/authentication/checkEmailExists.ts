import {createReducer} from "redux-create-reducer";
import produce from 'immer';
import {
    Action,
    asyncAction,
    AsyncAction,
    buildAction,
    buildSaga,
    buildSelector,
    ResolvedAsyncAction,
    startedAction,
    succeededAction
} from "../redux";
import {request} from "../api";

/** Status to indicate that we've started to check if an email address exists */
export const CHECK_STATUS_START = 'Started';

/** Status to indicate that the email address we've checked does exist */
export const CHECK_STATUS_EXISTS = 'Exists';

/** Status to indicate that the email address we've checked does not exist */
export const CHECK_STATUS_UNKNOWN = 'Unknown';

/** The shape of the state for this module */
export interface State {
    email?: string,
    state?: string,
}

/** The initial state for this module */
const INITIAL_STATE: State = {

};

/** Prefix for actions in this module */
const MODULE_PREFIX = 'CheckEmailExists/';

//////// Check If Email Exists

/** The action type for checking if an email exists */
const CHECK_EMAIL_ACTION = MODULE_PREFIX + 'checkEmailExists';

/** The action type for resetting the check for if an email exists */
const RESET_ACTION = MODULE_PREFIX + 'reset';

/** The shape of the action to check if an email exists */
type CheckEmailExistsAction = string;

/**
 * Build the Redux action to check if an email exists
 * @param email the email to check
 */
export function checkEmailExists(email: string): Action<CheckEmailExistsAction> {
    return buildAction(CHECK_EMAIL_ACTION, email);
}

/**
 * Build the Redux action to check if an email exists
 */
export function reset(): Action<null> {
    return buildAction(RESET_ACTION, null);
}

/** The shape of the response from the server on checking if an email address exists */
type CheckEmailExistsServiceResponse = {
    exists: boolean,
};

/**
 * Saga to check if an email address is already registered with the server
 * @param action the action to handle
 */
export function* checkEmailExistsSaga(action: Action<CheckEmailExistsAction>) : IterableIterator<any> {
    yield asyncAction(CHECK_EMAIL_ACTION, async (email: string) => {
        const response = await request<CheckEmailExistsServiceResponse>('/emails/{email}', {
            method: 'GET',
            urlParams: {
                email,
            }
        });

        return response.body.exists;
    }, action.payload);
}

/**
 * Reducer for when we start to check if an email address exists
 * @param state the state to update
 * @param action The action to handle
 */
export function CheckEmailExistsStartReducer(state: State, action: AsyncAction<string>) : State {
    return produce(state, (draft: State) => {
        draft.email = action.input;
        draft.state = CHECK_STATUS_START;
    });
}

/**
 * Reducer for when we start to check if an email address exists
 * @param state the state to update
 * @param action The action to handle
 */
export function CheckEmailExistsSuccessReducer(state: State, action: ResolvedAsyncAction<string, boolean>) : State {
    return produce(state, (draft: State) => {
        draft.state = action.payload ? CHECK_STATUS_EXISTS : CHECK_STATUS_UNKNOWN;
    });
}

/**
 * Reducer for when we reset the check if an email address exists
 * @param state the state to update
 */
export function ResetReducer(state: State) : State {
    return produce(state, (draft: State) => {
        delete draft.state;
        delete draft.email;
    });
}

/** Selector to get the state of the Check Email process */
export const selectCheckEmailStatus = buildSelector(['checkEmailExists'], (state: State) => state.state);

/** Selector to get the Email Address we are working with */
export const selectCheckEmailValue = buildSelector(['checkEmailExists'], (state: State) => state.email);

//////// The actual module definitions

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
    [startedAction(CHECK_EMAIL_ACTION)] : CheckEmailExistsStartReducer,
    [succeededAction(CHECK_EMAIL_ACTION)] : CheckEmailExistsSuccessReducer,
    [RESET_ACTION] : ResetReducer
});

/** The sagas for this module */
export const sagas = [
    buildSaga(CHECK_EMAIL_ACTION, checkEmailExistsSaga),
];

/** The actual module */
export default {
    CHECK_STATUS_START,
    CHECK_STATUS_EXISTS,
    CHECK_STATUS_UNKNOWN,

    checkEmailExists,
    reset,

    selectCheckEmailStatus,
    selectCheckEmailValue
}
