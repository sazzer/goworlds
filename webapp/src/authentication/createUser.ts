import {createReducer} from "redux-create-reducer";
import {Action, asyncAction, buildAction, buildSaga} from "../redux";

/** Prefix for actions in this module */
const MODULE_PREFIX = 'CreateUser/';

/** The shape of the state for this module */
export interface State {
}

/** The initial state for this module */
const INITIAL_STATE: State = {
};

//////// Selectors for getting details from the state

//////// Action for creating a new user

/** The action type for creating a new user */
const CREATE_USER_ACTION = MODULE_PREFIX + 'createUser';

/** The shape of the action to create a new user */
declare type CreateUserAction = {
    email: string,
    name: string,
    password: string,
};

/**
 * Build the Redux action to create a new user
 * @param email the email to user
 * @param name The name to use
 * @param password The password to use
 */
export function createUser(email: string, name: string, password: string): Action<CreateUserAction> {
    return buildAction(CREATE_USER_ACTION, {
        email,
        name,
        password,
    });
}

/** The shape of the response from the server on creating a user */
type CreateUserServiceResponse = {
};

/**
 * Saga to actually create a new user
 * @param action the action to handle
 */
export function* createUserSaga(action: Action<CreateUserAction>) : IterableIterator<any> {
    yield asyncAction(CREATE_USER_ACTION, async (email: string) => {
        return action;
    }, action.payload);
}

//////// The actual module definitions

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
});

/** The sagas for this module */
export const sagas = [
    buildSaga(CREATE_USER_ACTION, createUserSaga),
];

/** The actual module */
export default {
    createUser,
}
