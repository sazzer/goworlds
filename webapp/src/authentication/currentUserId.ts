import {createReducer} from "redux-create-reducer";
import {Action, BaseAction, buildAction} from "../redux";
import produce from "immer";

/** Prefix for actions in this module */
const MODULE_PREFIX = 'CurrentUser/';

/** The shape of the state for this module */
export declare type State = {
    userId: string | undefined,
};

/** The initial state for this module */
const INITIAL_STATE: State = {
    userId: undefined,
};

//////// Action for checking if an email address exists

/** The action type for storing the current User ID */
const STORE_CURRENT_USER_ACTION = MODULE_PREFIX + 'store';

/** The shape of the action to store the current User ID */
declare type StoreCurrentUserAction = {
    userId: string,
}

/**
 * Build the Redux action to store the current User ID
 * @param userId The User ID
 */
export function storeCurrentUser(userId: string): Action<StoreCurrentUserAction> {
    return buildAction(STORE_CURRENT_USER_ACTION, {
        userId
    });
}

/**
 * Reducer for storing the current User ID
 * @param state the state to update
 * @param action the action to update the state with
 * @return the new state
 */
export function storeCurrentUserReducer(state: State, action: Action<StoreCurrentUserAction>) : State {
    return produce(state, (draft: State) => {
        draft.userId = action.payload.userId;
    });
}

//////// The actual module definitions

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
    [STORE_CURRENT_USER_ACTION] : storeCurrentUserReducer as ((state: State, action: BaseAction) => State)
});
