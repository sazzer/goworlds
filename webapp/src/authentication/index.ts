import {createReducer} from "redux-create-reducer";
import produce from 'immer';
import {buildSaga, buildAction, Action, asyncAction} from "../redux";

/** The shape of the state for this module */
export type State = {

};

/** The initial state for this module */
const INITIAL_STATE: State = {

};

/** Prefix for actions in this module */
const MODULE_PREFIX = 'Authentication/';

//////// Check If Email Exists

const CHECK_EMAIL_ACTION = MODULE_PREFIX + 'checkEmailExists';

/** The shape of the action to check if an email exists */
type CheckEmailExistsAction = {
    email: string
};

/**
 * Build the Redux action to check if an email exists
 * @param email the email to check
 */
export function checkEmailExists(email: string): Action<CheckEmailExistsAction> {
    return buildAction(CHECK_EMAIL_ACTION, {
        email
    });
}

/**
 * Saga to check if an email address is already registered with the server
 * @param action the action to handle
 */
export function* checkEmailExistsSaga(action: Action<CheckEmailExistsAction>) : IterableIterator<any> {
    yield asyncAction(CHECK_EMAIL_ACTION, (email: string) => {
        return new Promise<string>((resolve) => {
            setTimeout(() => resolve('Success: ' + email), 5000);
        });
    }, action.payload.email);
}

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {

});

/** The sagas for this module */
export const sagas = [
    buildSaga(CHECK_EMAIL_ACTION, checkEmailExistsSaga),
];

/** The actual module */
export default {
    checkEmailExists
}
