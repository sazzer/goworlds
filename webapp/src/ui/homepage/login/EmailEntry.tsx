import React, {FunctionComponent, useState} from 'react';
import {Button, Form} from "semantic-ui-react";
import {useTranslation} from "react-i18next";
import {Formik} from "formik";
import * as Yup from 'yup';
import {FormikErrorMessage} from "../../common/FormikErrorMessage";
import {checkEmail} from "../../../authentication/checkEmail";
import {ErrorMessage} from "../../common/ErrorMessage";

/** The props that the EmailEntry area needs */
type EmailEntryProps = {
    onSubmit: (email: string, known: boolean) => void
};

/**
 * Component for allowing an email to be entered to start authentication
 * @constructor
 */
export const EmailEntry: FunctionComponent<EmailEntryProps> = ({onSubmit}) => {
    const { t } = useTranslation();
    const [submitting, setSubmitting] = useState<boolean>();
    const [submitError, setSubmitError] = useState<boolean>(false);

    const schema = Yup.object().shape({
        email: Yup.string()
            .required(t('loginArea.email.errors.required'))
            .email(t('loginArea.email.errors.email'))
    });

    const doSubmit = (email: string) => {
        setSubmitting(true);
        setSubmitError(false);

        checkEmail(email).subscribe(
            (status: boolean) => {
                setSubmitting(false);
                onSubmit(email, status);
            },
            (err: Error) => {
                console.log(err);
                setSubmitting(false);
                setSubmitError(true);
            }
        );
    };

    return (
        <Formik initialValues={{email: ''}}
                validationSchema={schema}
                onSubmit={(values) => doSubmit(values.email)}>
            {({values, isValid, errors, handleSubmit, handleChange, handleBlur}) =>
                <Form onSubmit={handleSubmit} error={!isValid || submitError} loading={submitting}>
                    <Form.Field required>
                        <label>
                            {t('loginArea.email.label')}
                        </label>
                        <Form.Input fluid
                                    placeholder={t('loginArea.email.placeholder')}
                                    name="email"
                                    type="text"
                                    value={values.email}
                                    error={errors.email !== undefined}
                                    onChange={handleChange}
                                    onBlur={handleBlur}
                                    autoFocus />
                        <FormikErrorMessage name="email" />
                    </Form.Field>
                    <Button type="submit" primary>
                        {t('loginArea.submit.loginRegister')}
                    </Button>
                    <ErrorMessage errors={[
                        submitError && t('loginArea.submit.errors.unexpected')
                    ]}/>
                </Form>
            }
        </Formik>
    );
};
