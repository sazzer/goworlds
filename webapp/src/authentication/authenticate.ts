import {
    Action,
    asyncAction,
    buildAction,
    buildSaga,
    failedAction,
    ResolvedAsyncAction,
    succeededAction
} from "../redux";
import {request, RequestError} from "../api";
import {put} from "redux-saga-test-plan/matchers";
import {storeAccessToken} from "./accessToken";
import jwtDecode from 'jwt-decode';
import {storeCurrentUser} from "./currentUserId";
import {MODULE_PREFIX} from "./module";
import {getConfig} from "../config";

//////// Action for authenticating to request an Access Token and ID Token

/** The action type for checking if an email exists */
const AUTHENTICATE_ACTION = MODULE_PREFIX + 'authenticate';

/** The shape of the action to authenticate a user */
declare type AuthenticateAction = {
    email: string,
    password: string,
    callback: (err: Error | undefined) => void
}

/**
 * Build the Redux action to authenticate a user
 * @param email the email to authenticate with
 * @param password The password to authenticate with
 * @param callback A callback to trigger when the action is completed
 */
export function authenticate(email: string, password: string, callback?: (err: Error | undefined) => void): Action<AuthenticateAction> {
    return buildAction(AUTHENTICATE_ACTION, {
        email,
        password,
        callback: callback || (() => {})
    });
}

/** The shape of the response from the server on requesting an access token */
declare type AuthenticationServiceResponse = {
    access_token: string,
    expires_in: number,
    token_type: string,
    id_token: string,
};

/** An Error from an OAuth2 Request */
export class OAuth2Error extends Error {
    /** The OAuth2 error code */
    errorCode: string;

    /**
     * Construct the error
     * @param error the error code
     */
    constructor(error: string) {
        super(error);

        this.errorCode = error;
    }
}

/**
 * Saga to authenticate a user
 * @param action the action to handle
 */
export function* authenticateSaga(action: Action<AuthenticateAction>) : IterableIterator<any> {
    yield asyncAction(AUTHENTICATE_ACTION, async () => {
        const requestBody = new FormData();
        requestBody.set('grant_type', 'tag:goworlds,2019:oauth2/grant_type/email_password');
        requestBody.set('email', action.payload.email);
        requestBody.set('password', action.payload.password);
        requestBody.set('scope', 'openid');

        try {
            const response = await request<AuthenticationServiceResponse>('/oauth2/token', {
                method: 'POST',
                body: requestBody,
                clientId: process.env.REACT_APP_API_CLIENTID || getConfig('API_CLIENTID'),
            });

            return response.body;
        } catch (err) {
            if (err instanceof RequestError) {
                throw new OAuth2Error(err.body.error);
            }
            throw err;
        }
    }, action.payload);
}

/**
 * Saga to trigger when Authentication fails for some reason
 * @param action the action to handle
 */
export function authenticateFailedSaga(action: ResolvedAsyncAction<AuthenticateAction, Error>) {
    if (action.input) {
        action.input.callback(action.payload);
    }
}

/**
 * Saga to trigger when Authentication succeeded
 * @param action the action to handle
 */
export function* authenticateSucceededSaga(action: ResolvedAsyncAction<AuthenticateAction, AuthenticationServiceResponse>) : IterableIterator<any> {
    if (action.payload) {
        const expiry = new Date(new Date().getTime() + (action.payload.expires_in * 1000));
        yield put(storeAccessToken(action.payload.access_token, expiry));

        const idToken = jwtDecode(action.payload.id_token) as any;
        if (idToken.sub) {
            yield put(storeCurrentUser(idToken.sub));
        }
    }

    if (action.input) {
        action.input.callback(undefined);
    }
}

//////// The actual module definitions

/** The sagas for this module */
export const sagas = [
    buildSaga(AUTHENTICATE_ACTION, authenticateSaga),
    buildSaga(succeededAction(AUTHENTICATE_ACTION), authenticateSucceededSaga),
    buildSaga(failedAction(AUTHENTICATE_ACTION), authenticateFailedSaga),
];
