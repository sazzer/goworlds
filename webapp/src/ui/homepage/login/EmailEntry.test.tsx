import React from 'react';
import {mount} from 'enzyme';
import {EMPTY, of, throwError} from "rxjs";
import {EmailEntry} from "./EmailEntry";
import * as mockCheckEmail from '../../../authentication/checkEmail';

jest.mock('../../../authentication/checkEmail');

/** Set up the component to test */
function setup() {
    const onSubmit = jest.fn();
    const element = mount(<EmailEntry onSubmit={onSubmit}/>);

    const submitForm = () => {
        return new Promise((resolved) => {
            element.find('form').simulate('submit', {
                preventDefault: () => {}
            });

            setImmediate(() => {
                element.update();
                resolved();
            });
        });
    };
    const enterEmail = (email) => element.find('input[name="email"]').simulate('change', {
        target: {
            name: 'email',
            value: email
        }
    });

    return {
        onSubmit,
        element,
        submitForm,
        enterEmail
    };
}

it('renders without crashing', () => {
    const {element} = setup();

    expect(element.render()).toMatchSnapshot();
});

it('renders the email address that was entered', () => {
    const {element, enterEmail} = setup();

    enterEmail('graham@grahamcox.co.uk');

    expect(element.render()).toMatchSnapshot();
});

it('submits the email address correctly', async () => {
    const {onSubmit, enterEmail, submitForm} = setup();

    (mockCheckEmail.checkEmail as jest.MockInstance).mockReturnValue(of(true));

    enterEmail('graham@grahamcox.co.uk');
    await submitForm();

    expect(onSubmit).toBeCalledTimes(1);
    expect(onSubmit).toBeCalledWith('graham@grahamcox.co.uk', true);
});

it('renders the form as loading whilst waiting for the API to return', async () => {
    const {element, enterEmail, submitForm} = setup();

    (mockCheckEmail.checkEmail as jest.MockInstance).mockReturnValue(EMPTY);

    enterEmail('graham@grahamcox.co.uk');
    await submitForm();

    expect(element.render()).toMatchSnapshot();
});

it("doesn't submit the email address if one wasn't provided", async () => {
    const {onSubmit, submitForm} = setup();

    await submitForm();

    expect(onSubmit).toBeCalledTimes(0);
});

it("renders an error if submitted without an email address", async () => {
    const {element, submitForm} = setup();

    await submitForm();

    expect(element.render()).toMatchSnapshot();
});

it("doesn't submit the email address if an API error occurred", async () => {
    const {onSubmit, enterEmail, submitForm} = setup();

    (mockCheckEmail.checkEmail as jest.MockInstance).mockReturnValue(throwError(new Error('Oops')));

    enterEmail('graham@grahamcox.co.uk');
    await submitForm();

    expect(onSubmit).toBeCalledTimes(0);
});


it("renders an error if an API error occurred", async () => {
    const {element, enterEmail, submitForm} = setup();

    (mockCheckEmail.checkEmail as jest.MockInstance).mockReturnValue(throwError(new Error('Oops')));

    enterEmail('graham@grahamcox.co.uk');
    await submitForm();

    expect(element.render()).toMatchSnapshot();
});
