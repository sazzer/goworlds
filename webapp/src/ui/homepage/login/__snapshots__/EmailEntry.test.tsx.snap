// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an error if an API error occurred 1`] = `
<form
  class="ui error form"
>
  <div
    class="required field"
  >
    <label>
      Email Address
    </label>
    <div
      class="field"
    >
      <div
        class="ui fluid input"
      >
        <input
          name="email"
          placeholder="Enter Email Address"
          type="text"
          value="graham@grahamcox.co.uk"
        />
      </div>
    </div>
  </div>
  <button
    class="ui primary button"
    type="submit"
  >
    Login/Register
  </button>
  <div
    class="ui error message"
  >
    An unexpected error occurred
  </div>
</form>
`;

exports[`renders an error if submitted without an email address 1`] = `
<form
  class="ui error form"
>
  <div
    class="required field"
  >
    <label>
      Email Address
    </label>
    <div
      class="error field"
    >
      <div
        class="ui fluid input"
      >
        <input
          name="email"
          placeholder="Enter Email Address"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="ui error message"
    >
      Please enter an Email Address
    </div>
  </div>
  <button
    class="ui primary button"
    type="submit"
  >
    Login/Register
  </button>
  <div
    class="ui error message"
  />
</form>
`;

exports[`renders the email address that was entered 1`] = `
<form
  class="ui form"
>
  <div
    class="required field"
  >
    <label>
      Email Address
    </label>
    <div
      class="field"
    >
      <div
        class="ui fluid input"
      >
        <input
          name="email"
          placeholder="Enter Email Address"
          type="text"
          value="graham@grahamcox.co.uk"
        />
      </div>
    </div>
  </div>
  <button
    class="ui primary button"
    type="submit"
  >
    Login/Register
  </button>
  <div
    class="ui error message"
  />
</form>
`;

exports[`renders the form as loading whilst waiting for the API to return 1`] = `
<form
  class="ui loading form"
>
  <div
    class="required field"
  >
    <label>
      Email Address
    </label>
    <div
      class="field"
    >
      <div
        class="ui fluid input"
      >
        <input
          name="email"
          placeholder="Enter Email Address"
          type="text"
          value="graham@grahamcox.co.uk"
        />
      </div>
    </div>
  </div>
  <button
    class="ui primary button"
    type="submit"
  >
    Login/Register
  </button>
  <div
    class="ui error message"
  />
</form>
`;

exports[`renders without crashing 1`] = `
<form
  class="ui error form"
>
  <div
    class="required field"
  >
    <label>
      Email Address
    </label>
    <div
      class="field"
    >
      <div
        class="ui fluid input"
      >
        <input
          name="email"
          placeholder="Enter Email Address"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <button
    class="ui primary button"
    type="submit"
  >
    Login/Register
  </button>
  <div
    class="ui error message"
  />
</form>
`;
